src/libertem/analysis/gridmatching.py

```diff
@@ -279,12 +279,14 @@ def full_match(cls, correlation_result: CorrelationResult, zero, cand=[], parame
                 match = match.weighted_optimize()
             except NotFoundException:
                 # print("no match found:\n", points)
+                unmatched = working_set
```

Comment:
I've added a test case with three residual points that tests one of these alternative exits. Testing the NotFoundException and the failure to refine separately would be difficult.

The first exit should be the common one. The second one is a safeguard to avoid an endless loop if a prototype match is found, but then refinement fails.